### PR TITLE
#377 SnackBar additions

### LINKED ISSUE #377: [Enhancement] Snackbar additions
## Summary

I am really looking forward to a flexible snackbar component in XCT. I gave it a quick spin but couldn't find any way to:

1) Set styling in a single place. I would like to set the styling for all snackbars (colors, fonts, default durations etc.) in one place. In fact outside of duration I couldn't find where to theme it at all, but I didn't really look too hard.

2) Couldn't find where to define placement (top, bottom etc.) and appearing direction and/or effect (fade in/out etc). But I didn't look too hard either.

3) More than one action. Sometimes you need more than one action, but the snackbar component only seems to allow adding one action.

4) Swipe to dismiss: Considering it seemed a little too bare bones to use atm I never tested my code with it, but swipe to dismiss is something users will expect. Is it supported? If not one will be forced to add a 'Close' action, which then will eat up the only action it allows right now.

5) This would be really really nice: Ability to add an image.

Thanks.

### PR DESCRIPTION
**Pull request by VladislavAntonyuk**
### Bugs Fixed ###

- Fixes #377 

### API Changes ###

Added: 
 
```csharp
public class SnackBarOptions : ActionOptions
	{
		public IEnumerable<SnackBarActionOptions> Actions { get; }
	}
```

```csharp
public class SnackBarActionOptions
	{
		public Func<Task> Action { get; set; }
		public string Text { get; set; }
		public double FontSize { get; set; } = 14;
		public string FontFamily { get; set; } = "Arial";
		public Color BackgroundColor { get; set; } = Color.White;
		public Color ForegroundColor { get; set; } = Color.Black;
	}
```

```csharp
public class ActionOptions
	{
		public MessageOptions MessageOptions { get; }
		public Color BackgroundColor { get; }
		public bool IsRtl { get; }
		public TaskCompletionSource<bool> Result { get; }
		public int Duration { get; }
		public void SetResult(bool result) => Result.TrySetResult(result);
	}
```

```csharp
public class MessageOptions
	{
		public string Message { get; set; }
		public double FontSize { get; set; } = 14;
		public string FontFamily { get; set; } = "Arial";
		public Color Color { get; set; } = Color.Default;
	}
```

Changed:

```csharp
public static class PageExtension
	{
		public static Task<bool> DisplayToastAsync(this Page page, string message, int duration = 3000)
		public static Task<bool> DisplaySnackBarAsync(this Page page, string message, string actionButtonText, Func<Task> action, int duration = 3000)
		public static Task<bool> DisplaySnackBarAsync(this Page page, SnackBarOptions arguments)
}
```

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Comment by VladislavAntonyuk** [comment]
UWP
image

Android
image

WPF
image

**Comment by AndreiMisiukevich** [comment]
@VladislavAntonyuk could you pls resolve merge conflicts

File: XamarinCommunityToolkit/Views/SnackBar/Options/SnackBarOptions.shared.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Collections.Generic;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.UI.Views.Options
+{
+	public class SnackBarOptions : ActionOptions
+	{
+		public SnackBarOptions(MessageOptions message, int duration, Color backgroundColor, bool isRtl, IEnumerable<SnackBarActionOptions> actions)
+			: base(message, duration, backgroundColor, isRtl) =>
+			Actions = actions ?? new List<SnackBarActionOptions>();
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/SnackBar/Options/SnackBarOptions.shared.cs:10]: ```Enumerable.Empty<SnackBarActionOptions>()``` ?

File: XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project Sdk="MSBuild.Sdk.Extras/2.0.54">
   <PropertyGroup>
-    <TargetFrameworks>netstandard1.0;netstandard2.0;netstandard2.1;Xamarin.iOS10;MonoAndroid90;MonoAndroid10.0;Xamarin.TVOS10;Xamarin.WatchOS10;Xamarin.Mac20;tizen40</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;netstandard2.1;Xamarin.iOS10;MonoAndroid90;MonoAndroid10.0;Xamarin.TVOS10;Xamarin.WatchOS10;Xamarin.Mac20;tizen40</TargetFrameworks>
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj:4]: I am not sure we want to drop support of ```netstandard1.0```
CC @jfversluis

File: XamarinCommunityToolkitSample/Pages/Views/ActionsPage.xaml.cs [hunk 3]
```diff
@@ -1,40 +1,83 @@
...
+using Xamarin.CommunityToolkit.UI.Views.Options;
 
 namespace Xamarin.CommunityToolkit.Sample.Pages.Views
 {
 	public partial class ActionsPage : BasePage
 	{
 		public ActionsPage() => InitializeComponent();
 
-		async void Button1_Clicked(object sender, EventArgs args)
+		async void Button1Clicked(object sender, EventArgs args)
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkitSample/Pages/Views/ActionsPage.xaml.cs:17]: can we choose more friendly names? :)

File: XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj [hunk 4]
```diff
@@ -75,12 +75,6 @@
     <None Include="..\LICENSE" PackagePath="" Pack="true" />
     <None Include="..\assets\XamarinCommunityToolkit_128x128.png" PackagePath="icon.png" Pack="true" />
   </ItemGroup>
-  <ItemGroup Condition=" $(TargetFramework.StartsWith('netstandard1.0')) ">
-    <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.5.4" />
-    <PackageReference Include="System.Reflection.Emit.Lightweight" Version="4.7.0" />
-    <PackageReference Include="System.ComponentModel" Version="4.3.0" />
-    <PackageReference Include="System.ValueTuple" Version="4.5.0" />
-  </ItemGroup>
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj:83]: Not sure we want to drop it... CC @jfversluis

**Review by AndreiMisiukevich — COMMENTED** [review]
Please wait until @jfversluis reviewed

File: XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj (hunk 2 above)
Comment (VladislavAntonyuk) [reply XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj:4]: netstandard1.0 doesn't support api: CultureInfo.Text.IsRightToLeft

File: XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj (hunk 4 above)
Comment (VladislavAntonyuk) [reply XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj]: reverted this changes

File: XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj (hunk 2 above)
Comment (AndreiMisiukevich) [reply XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj]: let's retrun it back as well)

File: XamarinCommunityToolkit/Views/SnackBar/Options/SnackBarOptions.shared.cs (hunk 1 above)
Comment (AndreiMisiukevich) [reply XamarinCommunityToolkit/Views/SnackBar/Options/SnackBarOptions.shared.cs]: Any concerns? :)

File: XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj (hunk 2 above)
Comment (VladislavAntonyuk) [reply XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj]: The file is outdated. I rolled back this changes

File: XamarinCommunityToolkit/Views/SnackBar/Options/SnackBarOptions.shared.cs (hunk 1 above)
Comment (VladislavAntonyuk) [reply XamarinCommunityToolkit/Views/SnackBar/Options/SnackBarOptions.shared.cs]: You are viewing outdated file.

**Review by AndreiMisiukevich — APPROVED** [review]

**Review by AndreiMisiukevich — APPROVED** [review]

File: XamarinCommunityToolkit/Views/SnackBar/Options/ActionOptions.shared.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
+﻿using System.Threading.Tasks;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.UI.Views.Options
+{
+	/// <summary>
+	/// Use for any actions: toasts, SnackBars etc.
+	/// </summary>
+	public class ActionOptions
```
Comment (jfversluis) [inline XamarinCommunityToolkit/Views/SnackBar/Options/ActionOptions.shared.cs:9]: Can we make this internal? It seems only the `SnackBarActionOptions` inheritance is used right now, right? If we find a use to actually allow users to use this object we can open it, but for now I would say keep it internal until we actually have use for it to be public.

**Comment by jfversluis** [comment]
Just one question I'd like to be clear on. Other than that, seems good

File: XamarinCommunityToolkit/Views/SnackBar/Options/ActionOptions.shared.cs (hunk 5 above)
Comment (VladislavAntonyuk) [reply XamarinCommunityToolkit/Views/SnackBar/Options/ActionOptions.shared.cs:9]: I added extension method ToastAsync
I will add it as a parameter.
Give me a minute

Comment (VladislavAntonyuk) [reply XamarinCommunityToolkit/Views/SnackBar/Options/ActionOptions.shared.cs:9]: done

**Comment by VladislavAntonyuk** [comment]
> Just one question I'd like to be clear on. Other than that, seems good

Added UTs and added new extension method for toasts with ActionOptions()

**Review by AndreiMisiukevich — APPROVED** [review]

**Comment by VladislavAntonyuk** [comment]
it is also partially fix https://github.com/xamarin/XamarinCommunityToolkit/issues/260

**Review by AndreiMisiukevich — APPROVED** [review]

**Pull request merged**